### PR TITLE
raftstore: wake up hibernated leader after failed read index

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

###  What have you changed?
Fix https://github.com/tikv/tikv/issues/6596
* Broadcast `MsgRegionWakeUp` to wake up potential hibernated leader when follower does not know who is the leader before requesting read index
* Change group state to `Ordered` after receiving a `MsgReadIndex` from other peers.

###  What is the type of the changes?

- Bugfix (a change which fixes an issue)

###  How is the PR tested?

- Unit test
- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No.

###  Does this PR affect `tidb-ansible`?

No.

### CODE REVIEW

**Comment by 5kbpers** [comment]
PTAL

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -392,6 +393,12 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         self.register_split_region_check_tick();
         self.register_check_peer_stale_state_tick();
         self.on_check_merge();
+        // Apply committed entries more quickly.
+        if self.fsm.peer.raft_group.get_store().committed_index()
```
Comment (gengliqi) [inline components/raftstore/src/store/fsm/peer.rs]: Maybe we can do the raft base tick instead?

Comment (5kbpers) [reply components/raftstore/src/store/fsm/peer.rs]: Seems the raft base tick does not generate `Ready` for followers.

**Comment by 5kbpers** [comment]
PTAL

File: components/raftstore/src/store/fsm/peer.rs [hunk 2]
```diff
@@ -835,7 +842,9 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
             || msg.get_message().get_msg_type() == MessageType::MsgTimeoutNow
         {
             self.reset_raft_tick(GroupState::Chaos);
-        } else if msg.get_from_peer().get_id() == self.fsm.peer.leader_id() {
+        } else if msg.get_from_peer().get_id() == self.fsm.peer.leader_id()
+            || msg.get_message().get_msg_type() == MessageType::MsgReadIndex
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs]: Seems covered by #6672

File: components/raftstore/src/store/peer.rs [hunk 3]
```diff
@@ -2006,6 +2011,15 @@ impl Peer {
...
             );
             poll_ctx.raft_metrics.invalid_proposal.read_index_no_leader += 1;
+            // The leader may be hibernated, send a message for trying to awaken the leader.
+            if poll_ctx.cfg.hibernate_regions
+                && (self.bcast_wake_up_time.is_none()
+                    || self.bcast_wake_up_time.as_ref().unwrap().elapsed()
+                        >= Duration::from_millis(MIN_BCAST_WAKE_UP_INTERVAL))
+            {
+                self.bcast_wake_up_message(&mut poll_ctx.trans);
+                self.bcast_wake_up_time = Some(Instant::now());
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs]: How about using `now_coarse`?

Comment (5kbpers) [reply components/raftstore/src/store/peer.rs]: For now `raftstore` only uses `std::time::Instant` rather `tikv_util::time::Instant`, we could repalce it later.

Comment (BusyJay) [reply components/raftstore/src/store/peer.rs]: Why not solve it now? I worry about the performance impact. Note that the timer is called not just for learner but also followers.

Comment (5kbpers) [reply components/raftstore/src/store/peer.rs]: Is there any risk to replace all of the `std::time::Instant` into `tikv_util::time::Instant`?

File: components/raftstore/src/store/fsm/peer.rs (hunk 1 above)
Comment (gengliqi) [reply components/raftstore/src/store/fsm/peer.rs]: I find a better place to set has_ready https://github.com/tikv/tikv/blob/7ae2018fe2213282796148af9f125804bf7abf35/components/raftstore/src/store/fsm/peer.rs#L1617.

**Pull request merged**